### PR TITLE
[GH-313] Add slash command to create Github Issue

### PR DESCRIPTION
**Pull request by GodlikePenguin**
#### Summary
Adds a new slash command to open the CreateIssue modal without needing to interact with a post.
Now we're able to create an issue without coming from a post we should ensure the server is able to handle issues being created with just a channelId rather than a postId.
If issues are created without a postId then don't append mattermost "created from" prompts to the Github issue

#### Ticket Link
  Fixes https://github.com/mattermost/mattermost-plugin-github/issues/313

### CODE REVIEW

**Comment by aaronrothschild** [comment]
This is great @GodlikePenguin thanks for your contribution on this, looking forward to using this myself!  :)

File: server/plugin/command.go [hunk 1]
```diff
@@ -572,6 +572,14 @@ func getAutocompleteData(config *Configuration) *model.AutocompleteData {
 	settings.AddStaticListArgument("", true, value)
 	github.AddCommand(settings)
 
+	issue := model.NewAutocompleteData("issue", "[command]", "Available commands: create")
+
+	issueCreate := model.NewAutocompleteData("create", "[description]", "Open a dialog to create a new issue in Github, using the description if provided")
+	issueCreate.AddTextArgument("Description for the Github issue", "[description]", "")
```
Comment (hanzei) [inline server/plugin/command.go:578]: Given that the title is the only required parameter, I'm wondering if the user should be able to input that instead of the description.

File: webapp/src/hooks/hooks.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import {openCreateIssueModalWithoutPost} from '../actions';
+
+const createIssueCommand = '/github issue create';
+
+export default class Hooks {
+    constructor(store) {
+        this.store = store;
+    }
+
+    slashCommandWillBePostedHook = (rawMessage, contextArgs) => {
```
Comment (hanzei) [inline webapp/src/hooks/hooks.js:13]: What was the reasoning behind doing the slash command handling in the webapp vs doing in server side? I seems weird to me to split the command handling between server and webapp.

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for the great PR :+1: 

One question to get the discussion going

File: webapp/src/hooks/hooks.js (hunk 2 above)
Comment (GodlikePenguin) [reply webapp/src/hooks/hooks.js:13]: I agree it's a little bit odd, this is how the implementation for Jira works too. Basically the problem here is that we want to open a modal when the comment comes in, and I was unable to see how to open the specific CreateIssue modal from the server mode. As far as I can see all the code for the modal is in the front end so unless I recreate from scratch (which would lead to having to maintain this modal in two places) then it's easiest to catch the command from the webapp and handle there.

File: server/plugin/command.go (hunk 1 above)
Comment (GodlikePenguin) [reply server/plugin/command.go:578]: I think it probably does make more sense for this to be title, you're more likely to want to note down something short in Mattermost and fill it in properly in the Github UI.

The reason I went for description here is that is how the Jira command is implemented, but considering these are separate commands, I don't think we have to follow that convention.

File: server/plugin/api.go [hunk 3]
```diff
@@ -1289,23 +1296,26 @@ func (p *Plugin) createIssue(w http.ResponseWriter, r *http.Request, userID stri
...
+	channelID := issue.ChannelID
+	message := fmt.Sprintf("Created GitHub issue [#%v](%v)", result.GetNumber(), result.GetHTMLURL())
+	if post != nil {
+		postID = post.Id
+		channelID = post.ChannelId
+		message += fmt.Sprintf(" from a [message](%s)", permalink)
 	}
 
-	message := fmt.Sprintf("Created GitHub issue [#%v](%v) from a [message](%s)", result.GetNumber(), result.GetHTMLURL(), permalink)
 	reply := &model.Post{
```
Comment (larkox) [inline server/plugin/api.go:1308]: Does it make sense to have this message be an ephemeral message, specially when it is not attached to a post?
CC @aaronrothschild

**Review by larkox — APPROVED** [review]
LGTM

File: webapp/src/hooks/hooks.js (hunk 2 above)
Comment (hanzei) [reply webapp/src/hooks/hooks.js:13]: I flow I had in mind was "Slash command" -> "websocket event to client" -> "client opens modal" -> "create issue request to plugin". Do you think this flow is reasonable? Can it be implemented in a DRY way?

File: server/plugin/api.go (hunk 3 above)
Comment (GodlikePenguin) [reply server/plugin/api.go:1308]: My reasoning was that it might be useful for other members in the channel to know that an issue had been created.

File: webapp/src/hooks/hooks.js (hunk 2 above)
Comment (GodlikePenguin) [reply webapp/src/hooks/hooks.js]: Just looked into how this would be implemented and it turned out to be pretty simple and I think does make more sense as the server remains the only place which parses and validates the commands. I've pushed these changes so hopefully it's now looking more as you would expect.

File: server/plugin/command.go (hunk 1 above)
Comment (GodlikePenguin) [reply server/plugin/command.go]: Made this change in the latest push.

**Comment by GodlikePenguin** [comment]
Any updates?

**Comment by hanzei** [comment]
@GodlikePenguin I'm a bit overwhelmed with review from hacktoberfest at the moment. Will try to get to your PR ASP.

File: server/plugin/api.go (hunk 3 above)
Comment (hanzei) [reply server/plugin/api.go:1308]: I agree with @larkox here. Posting the message in the channel seems unexpected to me. I would like to go with the conservative route and use an ephemeral post.

File: server/plugin/api.go [hunk 4]
```diff
@@ -1289,23 +1296,26 @@ func (p *Plugin) createIssue(w http.ResponseWriter, r *http.Request, userID stri
...
+		ChannelId: channelID,
+		RootId:    postID,
+		ParentId:  postID,
 		UserId:    userID,
 	}
 
 	_, appErr = p.API.CreatePost(reply)
 	if appErr != nil {
-		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: "failed to create notification post " + issue.PostID, StatusCode: http.StatusInternalServerError})
+		p.writeAPIError(w, &APIErrorResponse{ID: "", Message: "failed to create notification post, postID: " + issue.PostID + ", channelID: " + channelID, StatusCode: http.StatusInternalServerError})
```
Comment (hanzei) [inline server/plugin/api.go:1318]: While we are here: Could you please `LogWarn` the error?

File: server/plugin/api.go [hunk 5]
```diff
@@ -1289,23 +1296,26 @@ func (p *Plugin) createIssue(w http.ResponseWriter, r *http.Request, userID stri
...
 	}
 
-	message := fmt.Sprintf("Created GitHub issue [#%v](%v) from a [message](%s)", result.GetNumber(), result.GetHTMLURL(), permalink)
 	reply := &model.Post{
 		Message:   message,
-		ChannelId: post.ChannelId,
-		RootId:    rootID,
-		ParentId:  rootID,
+		ChannelId: channelID,
+		RootId:    postID,
```
Comment (hanzei) [inline server/plugin/api.go:1311]: Why was this changed?  Should we keep the old logic?

File: server/plugin/command.go [hunk 6]
```diff
@@ -483,7 +500,9 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
 
 	if f, ok := p.CommandHandlers[action]; ok {
 		message := f(c, args, parameters, info)
-		p.postCommandResponse(args, message)
+		if len(message) > 0 {
```
Comment (hanzei) [inline server/plugin/command.go:503]: Nit pick: It's idiomatic to use 
```suggestion
		if message != "" {
```
instead

File: server/plugin/api.go (hunk 5 above)
Comment (hanzei) [inline server/plugin/api.go:1311]: A bit of a nit pick: If a user triggers the slash command from the RHS, the response post in put in the center channel. It's not a big deal and can be solved later.

**Review by hanzei — CHANGES_REQUESTED** [review]
Awesome work :tada: 

A few comment below

**Comment by jfrerich** [comment]
Hi @GodlikePenguin and thanks for the PR!  Could you address the PR feedback from @hanzei and address the merge conflict?  Feel free to ask questions and for clarification as needed!

**Comment by GodlikePenguin** [comment]
Apologies I haven't had a chance to look at this recently, but will probably get another chance this weekend

**Comment by jasonblais** [comment]
Thanks @GodlikePenguin for the heads up! Appreciate your help through this contribution :)

**Comment by GodlikePenguin** [comment]
Apologies slightly messed up the git history and commits when I merged master back into my branch.

**Review by hanzei — APPROVED** [review]
Great work :tada: Thanks you very much @GodlikePenguin

**Comment by DHaussermann** [comment]
Hi @GodlikePenguin 
I've done a round of testing and I can see the slash command working.
 
While testing this feature I **was** seeing the CCS break and the webapp go blank. I wonder if it could possibly be related to this PR?

The error was at `<mmurl>/static/plugins/github/github_bfb942eaa868d56e_bundle.js`
Screen Shot 2020-12-03 at 5 29 09 PM

When I created an issue and then a few seconds after the ephemeral post has confirmed my issue creation I saw this error. I then had to refresh to see the webapp again. 

I noticed the sever was mistakenly on a pre-release build of `release-5.29` after updating to `5.29.1` I seem to no longer repro the issue but may just be intermittent so I figured I would share the error in case any more context can be provided.

cc @hanzei maybe you have some thoughts on this ☝️ 

Testing is ongoing.

**Comment by hanzei** [comment]
@DHaussermann That _seems_ unrelated. Could you please try to repo the issue in a dev environment? That way we would get a clearer error message.

**Comment by DHaussermann** [comment]
This does seem to occur when this build is deployed. I was mistaken, this issue is not intermittent for me. The CSS is breaking whenever I click the link to the GitHub issue. 

This happens both with issue created from a post or from the slash command without a post. I also happens when I click the link from a post made on a previous build. 

Server logs around this time show...
```
{"level":"error","ts":1607360938.6375718,"caller":"app/post.go:340","msg":"Failed to handle post events","error":"failed to save thread membership with postid=pppbek3a3igq3xj7smrrpymqiw userid=xu5bru8gztg6dqn6bp6qr1np1y: Error 1054: Unknown column 'UnreadMentions' in 'field list'","errorVerbose":"Error 1054: Unknown column 'UnreadMentions' in 'field list'\nfailed to save thread membership with postid=pppbek3a3igq3xj7smrrpymqiw userid=xu5bru8gztg6dqn6bp6qr1np1y\ngithub.com/mattermost/mattermost-server/v5/store/sqlstore.(*SqlThreadStore).SaveMembership\n\tgithub.com/mattermost/mattermost-server/v5/store/sqlstore/thread_store.go:122\ngithub.com/mattermost/mattermost-server/v5/store/sqlstore.(*SqlThreadStore).CreateMembershipIfNeeded\n\tgithub.com/mattermost/mattermost-server/v5/store/sqlstore/thread_store.go:191\ngithub.com/mattermost/mattermost-server/v5/store/retrylayer.(*RetryLayerThreadStore).CreateMembershipIfNeeded\n\tgithub.com/mattermost/mattermost-server/v5/store/retrylayer/retrylayer.go:6991\ngithub.com/mattermost/mattermost-server/v5/store/timerlayer.(*TimerLayerThreadStore).CreateMembershipIfNeeded\n\tgithub.com/mattermost/mattermost-server/v5/store/timerlayer/timerlayer.go:6878\ngithub.com/mattermost/mattermost-server/v5/app.(*App).handlePostEvents\n\tgithub.com/mattermost/mattermost-server/v5/app/post.go:447\ngithub.com/mattermost/mattermost-server/v5/app.(*App).CreatePost\n\tgithub.com/mattermost/mattermost-server/v5/app/post.go:339\ngithub.com/mattermost/mattermost-server/v5/app.(*App).CreatePostMissingChannel\n\tgithub.com/mattermost/mattermost-server/v5/app/post.go:107\ngithub.com/mattermost/mattermost-server/v5/app.(*PluginAPI).CreatePost\n\tgithub.com/mattermost/mattermost-server/v5/app/plugin_api.go:544\ngithub.com/mattermost/mattermost-server/v5/plugin.(*apiTimerLayer).CreatePost\n\tgithub.com/mattermost/mattermost-server/v5/plugin/api_timer_layer_generated.go:600\ngithub.com/mattermost/mattermost-server/v5/plugin.(*apiRPCServer).CreatePost\n\tgithub.com/mattermost/mattermost-server/v5/plugin/client_rpc_generated.go:2836\nreflect.Value.call\n\treflect/value.go:460\nreflect.Value.Call\n\treflect/value.go:321\nnet/rpc.(*service).call\n\tnet/rpc/server.go:377\nruntime.goexit\n\truntime/asm_amd64.s:1373"}
``` 
```
{"level":"warn","ts":1607360938.6306877,"caller":"app/notification.go:290","msg":"Failed to update thread autofollow from mention","post_id":"bcwdcb5w1tyrtj1frf5ukdx13r","channel_id":"8m7hk8tkm3bxdyuwi8cd9usrqc","error":"SendNotifications: Failed to update thread membership for mentioned user, failed to save thread membership with postid=pppbek3a3igq3xj7smrrpymqiw userid=xu5bru8gztg6dqn6bp6qr1np1y: Error 1054: Unknown column 'UnreadMentions' in 'field list'"}
{"level":"warn","ts":1607360938.6307623,"caller":"app/notification.go:290","msg":"Failed to update thread autofollow from mention","post_id":"bcwdcb5w1tyrtj1frf5ukdx13r","channel_id":"8m7hk8tkm3bxdyuwi8cd9usrqc","error":"SendNotifications: Failed to update thread membership for mentioned user, failed to save thread membership with postid=pppbek3a3igq3xj7smrrpymqiw userid=s1jrzjbtsfffzjdkkjeztfdyue: Error 1054: Unknown column 'UnreadMentions' in 'field list'"}
```

```
{"level":"warn","ts":1607360559.7282615,"caller":"mlog/sugar.go:27","msg":"Failed to create DM post","plugin_id":"github","userID":"s1jrzjbtsfffzjdkkjeztfdyue","error":"Post.IsValid: Invalid message., id=inshywyww7geupy76tddii8hua"}
```

Steps:
- Deploy this build
- Enable and configure GitHub
- Type `/github issue create CSS will break`
- Add info and finish creating your issue in the modal
- Note the post appears in the thread with a link to the issue in GitHub
- Click the link
**Observed** CSS breaks 

@GodlikePenguin  can you repro this?

**Comment by GodlikePenguin** [comment]
So from my test setup (https://hub.docker.com/r/mattermost/mattermost-preview with just this plugin installed) I can't find any issues when following links to Github issues (either posted ephemerally or as part of a thread). I also have no issues in the Javascript console.

**Comment by hanzei** [comment]
I can reproduce the issue that @DHaussermann found:
Screenshot from 2020-12-09 12-30-52

But the same issue also happens on `master` of this plugin on my local environment. Hence, I don't think it's a blocker for this PR and I would much rather file a ticket for investigation.

**Comment by DHaussermann** [comment]
@hanzei was correct this does not repo on the last release but doe repo on master.   This bug does repo in `master` I have created https://github.com/mattermost/mattermost-plugin-github/issues/412 to address this separately. Sorry for the confusion.

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- Creating an issue from slash command works both with and without passing in a title 
- /autocomplete has been updated to accommodate 
- Post created is ephemeral and shows expected info and link
- Regression tested using post menu option - No impact
- Created a separate Issue for CSS break

LGTM!

Updated release tests to include this feature. 

Thanks @GodlikePenguin for this enhancement!
